obspy/io/seg2/seg2.py

```diff
@@ -262,52 +285,54 @@ def parse_free_form(self, free_form_str, attrib_dict):
         attrib_dict.
         """
-        # Separate the strings.
-        strings = free_form_str.split(self.string_terminator)
-        # This is not fully according to the SEG-2 format specification (or
-        # rather the specification only speaks about on offset of 2 bytes
-        # between strings and a string_terminator between two free form
-        # strings. The file I have show the following separation between two
-        # strings: 'random offset byte', 'string_terminator',
-        # 'random offset byte'
-        # Therefore every string has to be at least 3 bytes wide to be
-        # acceptable after being split at the string terminator.
-
-        def is_good_char(c):
-            return c in (b'0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMN'
-                         b'OPQRSTUVWXYZ!"#$%&\'()*+,-./:; <=>?@[\\]^_`{|}~ ')
-
-        # A loop over a bytestring in Python 3 returns integers. This can be
-        # solved with a number of imports from the python-future module and
-        # all kinds of subtle changes throughout this file. Separating the
-        # handling for Python 2 and 3 seems the cleaner and simpler approach.
-        if PY2:
-            strings = ["".join(filter(is_good_char, _i))
-                       for _i in strings
-                       if len(_i) >= 3]
-        else:
-            strings = ["".join(map(chr, filter(is_good_char, _i)))
-                       for _i in strings
-                       if len(_i) >= 3]
+        def cleanup_and_decode_string(value):
+            # Some software/hardware produces invalid characters.
+            def is_good_char(c):
+                return c in (b'0123456789'
+                             b'abcdefghijklmnopqrstuvwxyz'
+                             b'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+                             b'!"#$%&\'()*+,-./:; <=>?@[\\]^_`{|}~ ')
+
+            # A loop over a bytestring in Python 3 returns integers. This can
+            # be solved with a number of imports from the python-future module
+            # and all kinds of subtle changes throughout this file. Separating
+            # the handling for Python 2 and 3 seems the cleaner and simpler
+            # approach.
+            if PY2:
+                return "".join(filter(is_good_char, value)).strip()
+            else:
+                return "".join(map(chr, filter(is_good_char, value))).strip()
+
+        # Separate the strings. Every string starts with a 2-byte offset to the
+        # next string, and ends with a terminator (which seems redundant). An
+        # offset of 0 indicates the end of the strings.
+        offset = 0
+        strings = []
+        while offset + 2 < len(free_form_str):
+            strlen, = unpack_from(self.endian + b'H', free_form_str, offset)
+            if strlen == 0:
+                break
+            strings.append(
+                free_form_str[offset + 2:offset + strlen].rstrip(
+                    self.string_terminator))
```

Comment:
Not sure if we want to split at first `string_terminator` and drop the rest instead of `rstrip`. This is something an in-place edit might do to avoid re-writing the file.

Reply:
I'm not entirely sure what you mean by the in-place edit. Other than that I imagine that `rstrip()` is faster than a split because it only has to create one new string object instead of at least two for the `split()` variant.

Reply:
SEG2 seems to be a format with intentional 'holes' to allow for in-place editing (e.g., the trace pointer block can be larger than the number of traces, so you can just append a trace and overwrite one pointer).

So what I'm saying here is, if you have something like `Some very long text string\0`, you could replace it by just overwriting the beginning with `A shorter string\0` and end up with `A shorter string\0t string\0`. In that case, only stripping the right give you the wrong thing. Writing this out has mostly convinced me we should do the split instead, but I seem to recall another format where we might have done differently.

Reply:
I guess it is a question if the 2-byte offset in front of each string of the terminator takes precedence. I'd personally vote for the offset but I have no further information.

Reply:
I've actually decided to switch it (though not using `split`). It's not exactly about precedence. The offset only tells you where the next string _starts_; it doesn't necessarily tell you where the current string _ends_.